### PR TITLE
[Enhancement](Wal)Support dynamic wal space limit

### PR DESCRIPTION
**Pull request by Yukang-Lian**
## Proposed changes

Issue Number: close #xxx

User settings come in two scenarios: the first involves a percentage of disk space, while the second entails a fixed value.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run pipelinex_p0

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/http/action/http_stream.cpp [hunk 1]
```diff
@@ -182,12 +183,22 @@ int HttpStreamAction::on_header(HttpRequest* req) {
...
-            st = Status::InternalError("Http load size too large.");
+            std::stringstream err_msg;
+            const auto& map = ExecEnv::GetInstance()->wal_mgr()->get_wal_dir_to_max_limit_map();
+            err_msg << "When enable group commit, the data size can't be too large. The data size "
+                       "for this http load("
+                    << (req->header(HttpHeaders::CONTENT_LENGTH).empty()
+                                ? 0
+                                : std::stol(req->header(HttpHeaders::CONTENT_LENGTH)))
+                    << " Bytes) exceeds the WAL (Write-Ahead Log) limit ("
+                    << std::max_element(map.begin(), map.end(),
```
Comment (mymeiyi) [inline be/src/http/action/http_stream.cpp:194]: It's better to abstract a function to get the max limit wal dir

File: be/src/olap/wal_writer.h [hunk 2]
```diff
@@ -34,34 +34,28 @@ extern const uint32_t k_wal_magic_length;
...
 
     Status init();
     Status finalize();
 
     Status append_blocks(const PBlockArray& blocks);
-    size_t disk_bytes() const { return _disk_bytes.load(std::memory_order_relaxed); };
     Status append_header(uint32_t version, std::string col_ids);
 
     std::string file_name() { return _file_name; };
+    static std::string get_base_path(const std::string& wal_path);
```
Comment (mymeiyi) [inline be/src/olap/wal_writer.h:47]: Is this function unused?

File: be/src/vec/sink/group_commit_block_sink.cpp [hunk 3]
```diff
@@ -200,9 +200,31 @@ Status GroupCommitBlockSink::_add_blocks() {
     load_id.__set_lo(_load_id.lo);
     if (_load_block_queue == nullptr) {
         if (_state->exec_env()->wal_mgr()->is_running()) {
+            std::string wal_dir;
+            auto* wal_mgr = ExecEnv::GetInstance()->wal_mgr();
+            {
+                std::unique_lock l(lock);
+                auto map = wal_mgr->get_wal_dir_to_disk_usage_map();
+                wal_dir = map.size() == 1
```
Comment (mymeiyi) [inline be/src/vec/sink/group_commit_block_sink.cpp:208]: Abstract the select wal_dir to a method.
Is it better to move this method into the while loop?

File: be/src/vec/sink/group_commit_block_sink.cpp [hunk 4]
```diff
@@ -200,9 +200,31 @@ Status GroupCommitBlockSink::_add_blocks() {
...
+                    ExecEnv::GetInstance()->group_commit_mgr()->cv.wait_for(
+                            l, std::chrono::milliseconds(1000));
+                }
+            }
+
             RETURN_IF_ERROR(_state->exec_env()->group_commit_mgr()->get_first_block_load_queue(
                     _db_id, _table_id, _base_schema_version, load_id, _load_block_queue,
-                    _state->be_exec_version()));
+                    _state->be_exec_version(), wal_dir));
+            RETURN_IF_ERROR(wal_mgr->add_wal_path(_db_id, _table_id, _load_block_queue->txn_id,
```
Comment (mymeiyi) [inline be/src/vec/sink/group_commit_block_sink.cpp:226]: can move this to `LoadBlockQueue::create_wal` to hide the implementation

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/http/action/http_stream.cpp [hunk 5]
```diff
@@ -169,25 +170,39 @@ int HttpStreamAction::on_header(HttpRequest* req) {
...
-        group_commit_mode = "";
-    }
     if (!group_commit_mode.empty() && !iequal(group_commit_mode, "sync_mode") &&
         !iequal(group_commit_mode, "async_mode") && !iequal(group_commit_mode, "off_mode")) {
         st = Status::InternalError("group_commit can only be [async_mode, sync_mode, off_mode]");
-        if (iequal(group_commit_mode, "off_mode")) {
-            group_commit_mode = "";
-        }
+    } else if (group_commit_mode.empty() || iequal(group_commit_mode, "off_mode")) {
+        group_commit_mode = "off_mode";
```
Comment (mymeiyi) [inline be/src/http/action/http_stream.cpp:177]: set group_commit_mode = "", otherwise it is confilct with line line 184

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Review by mymeiyi — APPROVED** [review]
LGTM

File: be/src/common/config.cpp [hunk 6]
```diff
@@ -1134,8 +1134,9 @@ DEFINE_String(default_tzfiles_path, "${DORIS_HOME}/zoneinfo");
 // Max size(bytes) of group commit queues, used for mem back pressure, defult 64M.
 DEFINE_Int32(group_commit_max_queue_size, "67108864");
 
-// Max size(bytes) of wal disk using, used for disk space back pressure, default 64M.
-DEFINE_Int32(wal_max_disk_size, "67108864");
+// Max size(bytes) or percentage(%) of wal disk usage, used for disk space back pressure, default 10% of the disk available space.
+// wal_max_disk_limit=1024 or wal_max_disk_limit=10% can be automatically identified.
+DEFINE_String(wal_max_disk_limit, "10%");
```
Comment (mymeiyi) [inline be/src/common/config.cpp:1139]: bettern rename it  `group_commit_wal_max_disk_limit`

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/wal_writer.h [hunk 7]
```diff
@@ -34,34 +34,27 @@ extern const uint32_t k_wal_magic_length;
...
-    std::shared_ptr<std::condition_variable> cv;
     static const int64_t VERSION_SIZE = 4;
 
 private:
     static constexpr size_t MAX_WAL_WRITE_WAIT_TIME = 1000;
     std::string _file_name;
     io::FileWriterPtr _file_writer;
-    std::atomic_size_t _disk_bytes;
-    std::shared_ptr<std::atomic_size_t> _all_wal_disk_bytes;
     std::mutex _mutex;
```
Comment (mymeiyi) [inline be/src/olap/wal_writer.h:57]: is this unsed?

**Comment by Yukang-Lian** [comment]
run buildall

**Review by mymeiyi — APPROVED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run external

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**